src/group_theory/group_algebra.lean

```diff
@@ -0,0 +1,274 @@
+`monoid_algebra`.
+
+Some of this file is done in terms of monoids for the sake of generality, but after future
+additions, the bulk of the file will be about groups.
+
+## Tags
+
+group algebra, group ring, group cohomology, monoid algebra
+-/
+
+noncomputable theory
+
+universes u₁ u₂ u₃
+
+/-- The group ring `k[G]` of a group `G` with coefficients in a ring `k` (although this is defined
+for any type `G` and any semiring `k`, inducing less algebraic structure). -/
+@[derive [inhabited, add_comm_monoid]] def group_algebra (k : Type*) [semiring k] (G : Type*) :=
+monoid_algebra k G
+
+namespace group_algebra
+
+variables {k : Type*} {G : Type*}
+
+open monoid_algebra
+
+section
+
+instance [semiring k] : has_coe_to_fun (group_algebra k G) (λ _, G → k) :=
+monoid_algebra.has_coe_to_fun k G
+
+instance [semiring k] [nontrivial k] [nonempty G] : nontrivial (group_algebra k G) :=
+monoid_algebra.nontrivial
+
+instance [semiring k] [mul_one_class G] : non_assoc_semiring (group_algebra k G) :=
+monoid_algebra.non_assoc_semiring
+
+instance [semiring k] [monoid G] : semiring (group_algebra k G) :=
+monoid_algebra.semiring
+
+instance [comm_semiring k] [comm_monoid G] : comm_semiring (group_algebra k G) :=
+monoid_algebra.comm_semiring
+
+instance [ring k] : add_comm_group (group_algebra k G) :=
+monoid_algebra.add_comm_group
+
+instance [ring k] [monoid G] : ring (group_algebra k G) :=
+monoid_algebra.ring
+
+instance [comm_ring k] [comm_monoid G] : comm_ring (group_algebra k G) :=
+monoid_algebra.comm_ring
+
+instance {R : Type*} [semiring k] [monoid R] [distrib_mul_action R k] :
+  has_scalar R (group_algebra k G) :=
+monoid_algebra.has_scalar
+
+instance {R : Type*} [semiring k] [monoid R] [distrib_mul_action R k] :
+  distrib_mul_action R (group_algebra k G) :=
+monoid_algebra.distrib_mul_action
+
+instance {R : Type*} [semiring k] [semiring R] [module R k] :
+  module R (group_algebra k G) :=
+finsupp.module G k
+
+instance {A : Type*} [comm_semiring k] [semiring A] [algebra k A] [monoid G] :
+  algebra k (group_algebra A G) :=
+monoid_algebra.algebra
+
+/-- The natural embedding of `G` into its group algebra `k[G]`. -/
+def of (k : Type*) (G : Type*) [semiring k] [mul_one_class G] :
+  G →* group_algebra k G :=
+monoid_algebra.of k G
+
+@[simp] lemma of_apply [semiring k] [mul_one_class G] (x : G) :
+  of k G x = finsupp.single x 1 := rfl
+
+lemma induction_on [semiring k] [monoid G] {p : group_algebra k G → Prop}
+  (f : group_algebra k G) (hM : ∀ g, p (of k G g))
+  (hadd : ∀ f g : group_algebra k G, p f → p g → p (f + g))
+  (hsmul : ∀ (r : k) f, p f → p (r • f)) : p f :=
+monoid_algebra.induction_on _ hM hadd hsmul
+
+end
+
+/-- Given a map `• : G → M → M` for `M` a `k`-module, the map `k[G] → M → M` sending
+`(Σ kᵢgᵢ, m) ↦ Σ kᵢ• (gᵢ • m)`. If this were an instance, it would conflict with the natural
+action of `k[k]` on itself given by multiplication. -/
```

Comment:
It could be an instance if `G` is a group and `k` is a nontrivial semiring, because nontrivial semirings can never be groups because of the 0. Maybe this sort of things is what makes monoid algebras difficult to work with? If this PR is supposed to be about group algebras, and anyone who is desperate to do things with monoid algebras can use `monoid_algebra`, then why not demand that `G` is a group in the definition of the typeclass?

Reply:
Eric points out that if `G` is syntactically `units k` then we end up with two actions of `G` on `k[G]`.

Reply:
I think this instance is actually the safe one, it's the `comap` one that worries me

Reply:
(I think the diamonds caused by this one are actually propositionally equal up to `finset.sum_comm`

Reply:
Am I alright to make this an instance then, and strengthen assumptions on `k` and `G`? Also, are either of these valid approaches to the `comap` instance: https://github.com/leanprover-community/mathlib/pull/11574#discussion_r819851714